File: Binary Tree/106.从中序与后序遍历序列构造二叉树.cpp
#include <vector>

using namespace std;
/*
 * @lc app=leetcode.cn id=106 lang=cpp
 *
 * [106] 从中序与后序遍历序列构造二叉树
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution
{
public:
    // 切割数组
    TreeNode *traversal(vector<int> &inorder, vector<int> &postorder)
    {
        if (postorder.size() == 0)
        {
            return nullptr;
        }

        // 后续的最后一个节点为根节点
        int middle = postorder.back();
        TreeNode *root = new TreeNode(middle);

        if (postorder.size() == 1)
        {
            return root;
        }
        // 后序遍历舍弃掉中间的节点
        postorder.pop_back();
        int index;
        int isLeft = true;
        vector<int> inOrderLeft;
        vector<int> inOrderRight;

        // 分割中序遍历数组
        for (index = 0; index < inorder.size(); index++)
        {
            if (inorder[index] == middle)
            {
                isLeft = false;
                continue;
            }

            if (isLeft)
            {
                inOrderLeft.push_back(inorder[index]);
            }
            else
            {
                inOrderRight.push_back(inorder[index]);
            }
        }

        //使用中序遍历的左边长度，切割后序遍历的数组
        vector<int> postOrderLeft(postorder.begin(), postorder.begin() + inOrderLeft.size());
        vector<int> postOrderRight(postorder.begin() + inOrderLeft.size(), postorder.end());

        root->left = traversal(inOrderLeft, postOrderLeft);
        root->right = traversal(inOrderRight, postOrderRight);

        return root;
    }

    TreeNode *buildTree(vector<int> &inorder, vector<int> &postorder)
    {
        if (inorder.size() == 0 || postorder.size() == 0)
        {
            return nullptr;
        }

        return traversal(inorder, postorder);
    }
};
// @lc code=end

File: Binary Tree/二叉树的修改与构造/106.从中序与后序遍历序列构造二叉树.ts
/*
 * @lc app=leetcode.cn id=106 lang=typescript
 *
 * [106] 从中序与后序遍历序列构造二叉树
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */
function buildTree(inorder: number[], postorder: number[]): TreeNode | null {
  if (postorder.length === 0) {
    return null
  }

  if (postorder.length === 1) {
    return new TreeNode(postorder[0])
  }

  // 后序遍历的最后一个元素，为tree的顶点
  let top = postorder.pop()
  // 将tree作为分割点进行切割
  let i = 0
  for (; i < inorder.length; i++) {
    // 找到分割的位置
    if (inorder[i] === top) {
      break
    }
  }
  // 切割点将中序数组，切割成，左子树中序和右子树中序
  // 注意top节点在中序数组中间下标为ide位置，所有需要i+1
  let leftInorder = inorder.slice(0, i)
  let rightInorder = inorder.slice(i + 1)


  // 由于两个数组院子同一棵树，所以左右子树长度(节点数量)一定相同
  // 直接使用相同长度切割后序数组，这里top节点刚才已经pop去除，所以不用i+1
  let leftPostorder = postorder.slice(0, i)
  let rightPostorder = postorder.slice(i)

  return new TreeNode(top, buildTree(leftInorder, leftPostorder), buildTree(rightInorder, rightPostorder))
};
// @lc code=end

